client/state/data-layer/extensions-middleware.js

```diff
@@ -1,101 +1,58 @@
 /**
- * External dependencies
- */
-import { toPairs, fromPairs } from 'lodash';
-
-/**
  * Internal dependencies
  */
-import { local, mergeHandlers } from './utils';
-
-const stripHandlersForKey = ( key, keyHandlers, keyHandlersToStrip ) => {
-	const newKeyHandlers = (
-		keyHandlersToStrip
-			? keyHandlers.filter( ( handler ) => ! keyHandlersToStrip.includes( handler ) )
-			: keyHandlers
+import { mergeHandlers } from './utils';
+import { middleware } from './wpcom-api-middleware';
+import debugFactory from 'debug';
+
+const debug = debugFactory( 'calypso:state:data-layer' );
+let extensionHandlers = Object.create( null );
+let handlerMiddleware = buildMiddleware( extensionHandlers );
+
+export function addHandlers( extensionName, handlers, existingHandlers = extensionHandlers ) {
+	if ( Object.keys( existingHandlers ).includes( extensionName ) ) {
+		throw new Error( `Handlers for extension ${ extensionName } already present.` );
+	}
+
+	// Set the module globals for convenience
+	extensionHandlers = { ...existingHandlers, [ extensionName ]: handlers };
+	handlerMiddleware = buildMiddleware( extensionHandlers );
+
+	return extensionHandlers;
+}
+
+export function removeHandlers( extensionName, existingHandlers = extensionHandlers ) {
+	const { [ extensionName ]: omitted, ...remainingHandlers } = existingHandlers;
+
+	if ( ! omitted ) {
+		debug( `Trying to remove nonexistent handlers for ${ extensionName }` );
+	}
+
+	// Set the module globals for convenience
+	extensionHandlers = remainingHandlers;
+	handlerMiddleware = buildMiddleware( extensionHandlers );
+
+	return extensionHandlers;
+}
+
+export function buildMiddleware( handlersByExtension ) {
+	const allHandlers = Object.values( handlersByExtension ).reduce(
+		( accumulated, handlers ) => {
+			return mergeHandlers( accumulated, handlers );
+		},
```

Comment:
Gotcha. Curious about your usage of `Object.create( null )` and it's benefits over `{}`.